test/contrib/automotive/xcp/xcp.uts

```diff
@@ -0,0 +1,823 @@
+assert program_start_response.queue_size_pgm == 0xdd
+
+= test_program_clear(self):
+conf.contribs['XCP']['byte_order'] = 0
+
+cto_request = CTORequest(pid="PROGRAM_CLEAR") / ProgramClear(b'\x00\xff\xff\x00\x01\x00\x00')
+assert cto_request.pid == 0xd1
+assert bytes(cto_request) == b'\xd1\x00\xff\xff\x00\x01\x00\x00'
+
+assert cto_request['ProgramClear'].mode == 0x00
+assert cto_request['ProgramClear'].clear_range == 0x00000100
+
+cto_response = CTOResponse(b'\xff')
+assert cto_response.packet_code == 0xFF
+assert cto_response.answers(cto_request)
+
+= test_program
+
+conf.contribs['XCP']['byte_order'] = 0
+conf.contribs['XCP']['MAX_CTO'] = 8
+conf.contribs['XCP']['Address_Granularity_Byte'] = 1
+
+cto_request = CTORequest(pid="PROGRAM") / Program(b'\x06\x00\x01\x02\x03\x04\x05')
+assert cto_request.pid == 0xd0
+assert bytes(cto_request) == b'\xd0\x06\x00\x01\x02\x03\x04\x05'
+
+assert cto_request['Program'].nr_data_element == 0x06
+assert cto_request['Program'].data_elements == b"\x00\x01\x02\x03\x04\x05"
+
+cto_response = CTOResponse(b'\xff')
+assert cto_response.packet_code == 0xFF
+assert cto_response.answers(cto_request)
+
+
++ Tests on a virtual CAN-Bus
+= Imports
+
+import threading
+import time
+from subprocess import call
+
+import six
+
+from scapy.consts import LINUX
+from scapy.contrib.automotive.xcp.cto_commands_master import Connect, TransportLayerCmd, TransportLayerCmdGetSlaveId
+from scapy.contrib.automotive.xcp.scanner import XCPOnCANScanner
+from scapy.contrib.automotive.xcp.xcp import CTORequest, XCPOnCAN, CTOResponse, ConnectPositiveResponse, TransportLayerCmdGetSlaveIdResponse, GenericResponse
+from scapy.contrib.cansocket_python_can import CANSocket
+from scapy.main import load_layer
+
+= Load module
+
+load_layer("can")
+load_contrib("automotive.xcp.xcp",  globals_dict=globals())
+
+= Global variables
+
+iface0 = "vcan0"
+
+= Initialize a virtual CAN interface
+
+def set_up_virtual_python_can_socket(iface):
+    return CANSocket(bustype='virtual', channel=iface, timeout=0.01)
+
+print('setting up CAN')
+
+if 0 != call(["cansend", iface0,  "000#"]):
+    # vcan0 is not enabled
+    if 0 != call(["sudo", "modprobe", "vcan"]):
+        raise Exception("modprobe vcan failed")
+    if 0 != call(["sudo", "ip", "link", "add", "name", iface0, "type", "vcan"]):
+        print("add %s failed: Maybe it was already up?" % iface0)
+    if 0 != call(["sudo", "ip", "link", "set", "dev", iface0, "up"]):
+        raise Exception("could not bring up %s" % iface0)
+
+if 0 != call(["cansend", iface0,  "000#"]):
+    raise Exception("cansend doesn't work")
+
+print("CAN should work now")
+
+new_can_socket0 = lambda: CANSocket(bustype='virtual', channel=iface0, timeout=0.01)
+
+if six.PY3 and LINUX:
+    new_can_socket0 = lambda: CANSocket(iface0)
+
+if "python_can" in CANSocket.__module__:
+    new_can_socket0 = lambda: CANSocket(bustype='socketcan', channel=iface0, bitrate=250000, timeout=0.01)
+
```

Comment:
I suggest using:

```python
if LINUX:
    if six.PY3:
        new_can_socket0 = lambda: CANSocket(iface0)
    else:
        new_can_socket0 = lambda: CANSocket(bustype='socketcan', channel=iface0, bitrate=250000, timeout=0.01)
else:
    new_can_socket0 = lambda: CANSocket(bustype='virtual', channel=iface0, timeout=0.01)
```

I think this way the `linux` keyword above is not necessary anymore since it can run now on every platform
elif